gel/_internal/_deferred_import.py

```diff
@@ -0,0 +1,197 @@
+    runtime type evaluation.
+    """
+
+    def __init__(
+        self,
+        module: str,
+        attr: str | None = None,
+        alias: str | None = None,
+        package: str | None = None,
+    ) -> None:
+        """Initialize a DeferredImport proxy.
+
+        Args:
+            module: The module name to import. Can be a dotted name
+                like "os.path".
+            attr: Optional submodule name to import from the module.
+                If provided, this will import the specific submodule rather
+                than the entire module. Note: This parameter is only for
+                importing submodules, not attributes like classes or functions.
+            alias: Optional alias name to bind the imported object to in the
+                caller's namespace. If not provided, defaults to:
+                - The `attr` name if `attr` is specified
+                - The last component of the module name otherwise
+            package: Optional package name for relative imports. Used as the
+                anchor for resolving relative module names.
+
+        Examples:
+            >>> # Equivalent to: import os
+            >>> os = DeferredImport('os')
+
+            >>> # Equivalent to: from os import path (path is a submodule)
+            >>> path = DeferredImport('os', attr='path')
+
+            >>> # Equivalent to: import os.path as pathname
+            >>> pathname = DeferredImport('os.path', alias='pathname')
+
+            >>> # Equivalent to: from .parent import submodule
+            >>> submodule = DeferredImport('parent', attr='submodule',
+                                          package=__name__)
+
+        Note:
+            The actual import is deferred until the first attribute access on
+            the proxy object. Upon successful import, the imported object is
+            automatically bound to the caller's module namespace using the
+            determined alias name.
+
+            This class is designed specifically for importing modules and
+            submodules only, not individual attributes like classes or
+            functions.
+        """
+        if not alias:
+            if attr is not None:
+                alias = attr
+            else:
+                _, _, alias = module.rpartition(".")
+
+        self.__lm_import__ = _Import(
+            module=module,
+            attr=attr,
+            alias=alias,
+            package=package,
+        )
+        self.__lm_module__: weakref.ref[types.ModuleType] | None = None
+        self.__lm_importing_module__: weakref.ref[types.ModuleType] | None
+        if (caller_module := _get_caller_module()) is not None:
+            self.__lm_importing_module__ = weakref.ref(caller_module)
+        else:
+            self.__lm_importing_module__ = None
+
+    def __load(self) -> types.ModuleType:
+        if (mod_ref := self.__lm_module__) is not None and (
+            mod := mod_ref()
+        ) is not None:
+            return mod
```

Comment:
Should you replace itself with the resolved object once the real import is done?

Reply:
This is literally done a few lines down.  If we couldn't write to caller's globals when the actual import happened then we don't try again.

Reply:
ok I've missed it, cool